### DISCUSSION TITLE
Revisit idea of full server reboot on change

### DISCUSSION THREAD

**Discussion by kentcdodds** [remix-run/remix#6269] [General]
I'm excited about the cool stuff happening with HMR/HDR and the new changes in the Remix dev server in 1.16.0, but I'm really concerned about the apparent requirement that Remix reboot my server after every rebuild.

I have a few use cases where rebooting my server regularly impacts my workflow:

1. Starting up the server takes at least 500ms, sometimes longer. Mostly due to database connection time.
2. It becomes impossible for in-memory caches to persist between code changes
3. Chatty startup output

In some ways, number 2 is actually good because using `global.__whatever` stuff for persisting those in-memory caches is super annoying, but in other ways, not having those in-memory caches persisted between code changes is even more annoying. I suppose a solution could be to just use a persisted cache in a file system or something. Just takes a different approach.

But the startup time is definitely an issue. I'm guessing that HMR will be plenty fast when changing UI code, but HDR will always be as slow as it is to reboot the server which is not ideal. Maybe the solution is to simply improve startup time? 😅

For the chatty startup output, I like having the output of my server be descriptive (as I have contributors and workshop attendees). Here's an example of my output:

```
🔶 Mock server installed
🚀  We have liftoff!
Local:            http://localhost:3000
On Your Network:  http://172.20.20.20:3000
Press Ctrl+C to stop
```

Having all that output for every code change would be pretty annoying. I'm open to ideas though.

Can anyone clear things up for me a bit?

EDIT: After thinking about this a bit more, I think this is ok. But I added a third issue.

EDIT 2: @pcattori and I live streamed some of these topics: https://www.youtube.com/watch?v=IjE18rXpp9Q

**Comment by machour** [comment]
1 - I'm not fond of the full server reboot either, especially on low specs computers where it can take 2-3 seconds.

From my understanding, some server-side changes require a reboot (a change in a Prisma schema for example), while other just needs to revalidate all loaders.

Would it be possible to allow the developer to quickly switch from one more to another, using DevTools, or a debug bar ? 

<img width="1357" alt="Screenshot 2023-05-02 at 12 09 55 AM" src="https://user-images.githubusercontent.com/304450/235546963-c98a2f2b-6f1a-47c7-8c22-9b14b6a53cf5.png">

**Comment by pcattori** [comment]
For the chatty output: totally agree. Prettifying the logs and making them more concise is part of the work we want to do before stabilizing as `v2_dev`

For restarts: The `--no-restart` flag already exists for this purpose. It opts out of restarts and let's you manage server updates in your app server however you want. Note that you will be responsible for calling `broadcastDevReady` when your app server is initially up and running, as well as just after you purge the require cache. Check the `--no-restart` section of the [1.16 release notes for more](https://github.com/remix-run/remix/releases/tag/remix%401.16.0).
